include/swift/SIL/SILInstruction.h

```diff
@@ -5881,6 +5881,66 @@ class ObjCSuperMethodInst
       : UnaryInstructionBase(DebugLoc, Operand, Ty, Member) {}
 };
 
+// This is either the extracted method or the adjusted this ptr that
+// is the result of a CXXVirtualMethodInst.
+class CXXVirtualMethodInst;
+class CXXVirtualMethodResult final : public MultipleValueInstructionResult {
+public:
+  CXXVirtualMethodResult(unsigned index, SILType type,
+                         ValueOwnershipKind ownershipKind)
+      : MultipleValueInstructionResult(ValueKind::CXXVirtualMethodResult, index,
+                                       type, ownershipKind) {}
+
+  CXXVirtualMethodInst *getParent(); // inline below
+  const CXXVirtualMethodInst *getParent() const {
+    return const_cast<CXXVirtualMethodResult *>(this)->getParent();
+  }
+
+  bool isExtractedMethod() const { return getIndex() == 0; }
+  bool isAdjustedThisPtr() const { return getIndex() == 1; }
+
+  static bool classof(const SILNode *N) {
+    return N->getKind() == SILNodeKind::CXXVirtualMethodResult;
+  }
+};
+
+/// CXXVirtualMethodInst - Represents looking up a virtual method from a c++
+/// virtual method table and adjusting the this pointer.
+class CXXVirtualMethodInst final
+    : public UnaryInstructionBase<SILInstructionKind::CXXVirtualMethodInst,
+                                  MultipleValueInstruction>,
+      public MultipleValueInstructionTrailingObjects<CXXVirtualMethodInst,
+                                                     CXXVirtualMethodResult> {
```

Comment:
I wonder if there's a better way to have exactly two results than this. My SIL isn't strong enough. @gottesmm?